File: ts/src/data_structures/linked_lists/singly.ts
import type { LinkedList } from "./linked_list.interface.js";
import { ListNode as Node } from "./node.js";

export class SinglyLinkedList<T> implements LinkedList<T>, Iterable<T> {
    private sentinel: Node<T | null>
    private tail: Node<T>
    private length: number

    constructor() {
        this.sentinel = new Node<T | null>(null);
        this.tail = this.sentinel as Node<T>;
        this.length = 0;
    }

    size(): number {
        return this.length;
    }

    isEmpty(): boolean {
        return this.length === 0;
    }

    getAt(index: number): T {
        if (index < 0 || index >= this.length) {
            throw new RangeError("Index out of bounds");
        }

        let curr = this.sentinel.next as Node<T>;
        for (let i = 0; i < index; i++) {
            curr = curr.next as Node<T>;
        }

        return curr.value;
    }

    find(value: T): Node<T> | null {
        let curr = this.sentinel.next as Node<T>;
        while (curr) {
            if (curr.value === value) {
                return curr;
            }
            curr = curr.next as Node<T>;
        }
        return null;
    }

    append(value: T): void {
        const newNode = new Node<T>(value);
        this.tail.next = newNode;
        this.tail = newNode;
        this.length++;
    }

    prepend(value: T): void {
        const newNode = new Node<T>(value, this.sentinel.next as Node<T> | null);
        this.sentinel.next = newNode;
        if (this.length === 0) {
            this.tail = newNode;
        }
        this.length++;
    }

    insertAt(value: T, index: number): void {
        if (index < 0 || index > this.length) {
            throw new RangeError("Index out of bounds");
        }

        if (index === 0) {
            this.prepend(value);
            return;
        } else if (index === this.length - 1) {
            this.append(value);
        } else {
            let curr = this.sentinel as Node<T>;
            for (let i = 0; i <= index; i++) {
                curr = curr.next as Node<T>;
            }

            const newNode = new Node<T>(value, curr.next);
            curr.next = newNode;
            this.length++;
        }
    }

    deleteHead(): T | null {
        if (this.length === 0) {
            return null;
        }

        const head = this.sentinel.next as Node<T>;
        this.sentinel.next = head.next;

        this.length--;
        if (this.length === 0) {
            this.tail = this.sentinel as Node<T>;
        }

        return head.value;
    }

    deleteTail(): T | null {
        if (this.length === 0) {
            return null;
        }

        let curr = this.sentinel as Node<T>;
        while (curr.next !== this.tail) {
            curr = curr.next as Node<T>;
        }

        const tailValue = this.tail.value;
        curr.next = null;
        this.tail = curr;
        this.length--;

        return tailValue;
    }

    deleteAt(index: number): T | null {
        if (index < 0 || index >= this.length) {
            throw new RangeError("Index out of bounds");
        }

        if (index === 0) {
            return this.deleteHead();
        } else if (index === this.length - 1) {
            return this.deleteTail();
        } else {
            let curr = this.sentinel as Node<T>;
            for (let i = 0; i < index; i++) {
                curr = curr.next as Node<T>;
            }

            const nodeToDelete = curr.next as Node<T>;
            curr.next = nodeToDelete.next;
            this.length--;

            return nodeToDelete.value;
        }
    }

    delete(value: T): T | null {
        let curr = this.sentinel as Node<T>;

        for (let i = 0; curr.next; i++) {
            if (curr.next.value === value) {
                if (i == 0) {
                    return this.deleteHead();
                } else if (i == this.length - 1) {
                    return this.deleteTail();
                } else {
                    const nodeToDelete = curr.next as Node<T>;
                    curr.next = nodeToDelete.next;
                    this.length--;

                    return nodeToDelete.value;
                }
            }
        }

        return null;
    }

    clear(): void {
        while (this.length > 0) {
            this.deleteHead();
        }
    }

    merge(other: LinkedList<T>): void {
        for (const value of other) {
            this.append(value);
        }
    }

    reverse(): void {
        let prev: Node<T> | null = null;
        let curr = this.sentinel.next as Node<T> | null;
        this.tail = curr as Node<T>;

        while (curr) {
            const next = curr.next as Node<T> | null;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        this.sentinel.next = prev;
    }

    toArray(): T[] {
        let arr: T[] = [];

        let curr = this.sentinel.next as Node<T>;
        while (curr) {
            arr.push(curr.value);
            curr = curr.next as Node<T>;
        }

        return arr;
    }

    *[Symbol.iterator](): Iterator<T> {
        let curr = this.sentinel.next as Node<T>;
        while (curr) {
            yield curr.value;
            curr = curr.next as Node<T>;
        }
    }
}

File: python/src/dsa/data_structures/linked_lists/singly.py
from typing import Generic, Iterator, Self, TypeVar

from dsa.data_structures.linked_lists.interface import LinkedList

T = TypeVar("T")


class Node(Generic[T]):
    def __init__(self, value: T | None, next: Self | None = None) -> None:
        self.value = value
        self.next = next


class SinglyLinked(LinkedList, Generic[T]):
    def __init__(self) -> None:
        sentinel = Node(None)
        self.sentinel: Node[T] = sentinel
        self.tail: Node[T] = sentinel
        self.length: int = 0

    def __len__(self) -> int:
        return self.length

    def is_empty(self) -> bool:
        return self.length == 0

    def front(self) -> T | None:
        return self.sentinel.next.value if self.sentinel.next is not None else None

    def back(self) -> T | None:
        return self.tail.value if self.tail is not None else None

    def __getitem__(self, index: int) -> T:
        if index < 0 or index >= self.length:
            raise IndexError("index out of bounds")

        curr = self.sentinel.next
        for _ in range(index):
            assert curr is not None
            curr = curr.next

        assert curr is not None and curr.value is not None
        return curr.value

    def find(self, value: T) -> int:
        curr = self.sentinel.next
        idx = 0
        while curr is not None:
            assert curr.value is not None
            if curr.value == value:
                return idx

            idx += 1
            curr = curr.next

        return -1

    def prepend(self, value: T) -> None:
        new_node = Node(value, self.sentinel.next)
        self.sentinel.next = new_node

        if self.length == 0:
            self.tail = new_node

        self.length += 1

    def append(self, value: T) -> None:
        new_node = Node(value)
        self.tail.next = new_node
        self.tail = new_node
        self.length += 1

    def insert_at(self, index: int, value: T) -> None:
        if index < 0 or self.length > index:
            raise IndexError("index out of bounds")

        if index == 0:
            self.prepend(value)
        elif index == self.length:
            self.append(value)
        else:
            curr = self.sentinel
            for _ in range(index):
                assert curr is not None
                curr = curr.next

            assert curr is not None
            new_node = Node(value, curr.next)
            curr.next = new_node

            self.length += 1

    def __setitem__(self, index: int, value: T) -> None:
        if index < 0 or self.length >= index:
            raise IndexError("index out of bounds")

        curr = self.sentinel.next
        for _ in range(index):
            assert curr is not None
            curr = curr.next

        assert curr is not None
        curr.value = value

    def pop_front(self) -> T:
        node_to_delete = self.sentinel.next
        assert node_to_delete is not None

        self.sentinel.next = node_to_delete.next
        self.length -= 1

        assert node_to_delete.value is not None
        return node_to_delete.value

    def pop_back(self) -> T:
        curr = self.sentinel
        while curr.next is not None and curr.next != self.tail:
            curr = curr.next

        prev_tail_value = self.tail.value

        self.tail = curr
        self.length -= 1

        assert prev_tail_value is not None
        return prev_tail_value

    def remove_at(self, index: int) -> T:
        if index < 0 or self.length > index:
            raise IndexError("index out of bounds")

        if index == 0:
            return self.pop_front()
        elif index == self.length - 1:
            return self.pop_back()
        else:
            curr = self.sentinel
            for _ in range(index):
                assert curr is not None
                curr = curr.next

            assert curr is not None
            node_to_delete = curr.next

            assert node_to_delete is not None
            curr.next = node_to_delete.next

            self.length -= 1

            assert node_to_delete.value is not None
            return node_to_delete.value

    def __iter__(self) -> Iterator[T]:
        curr = self.sentinel.next
        while curr is not None:
            assert curr.value is not None
            yield curr.value
            curr = curr.next

    def __str__(self) -> str:
        list_str: list[str] = []

        curr = self.sentinel.next
        while curr is not None:
            assert curr.value is not None
            list_str.append(str(curr.value))

        return "[ " + " -> ".join(list_str) + " ]"


_ = SinglyLinked()
